tests/test_cmab_client.py

```diff
@@ -0,0 +1,161 @@
+
+    @patch("time.sleep", return_value=None)
+    def test_do_fetch_with_retry_exhausts_all_attempts(self, _):
+        failure_response = MagicMock()
+        failure_response.status_code = 500
+
+        self.mock_http_client.post.return_value = failure_response
+
+        result = self.client._do_fetch_with_retry("http://fake-url", {}, self.retry_config)
+        self.assertIsNone(result)
+        self.assertEqual(self.mock_http_client.post.call_count, 3)  # 1 original + 2 retries
+        self.mock_logger.error.assert_called_with(
+            Errors.CMAB_FETCH_FAILED.format("Exhausted all retries for CMAB request."))
+
+
+class TestDefaultCmabClientFetchDecision(unittest.TestCase):
+    def setUp(self):
+        self.mock_http_client = MagicMock()
+        self.mock_logger = MagicMock()
+        self.retry_config = CmabRetryConfig(max_retries=2, initial_backoff=0.01, max_backoff=1, backoff_multiplier=2)
+        self.client = DefaultCmabClient(
+            http_client=self.mock_http_client,
+            logger=self.mock_logger,
+            retry_config=self.retry_config
+        )
+        self.rule_id = 'test_rule'
+        self.user_id = 'user123'
+        self.attributes = {'attr1': 'value1'}
+        self.cmab_uuid = 'uuid-1234'
+
+    @patch.object(DefaultCmabClient, '_do_fetch', return_value='var-abc')
+    def test_fetch_decision_success_no_retry(self, mock_do_fetch):
+        result = self.client.fetch_decision(self.rule_id, self.user_id, self.attributes, self.cmab_uuid)
+        self.assertEqual(result, 'var-abc')
+        mock_do_fetch.assert_called_once()
+
+    @patch.object(DefaultCmabClient, '_do_fetch_with_retry', return_value='var-xyz')
+    def test_fetch_decision_success_with_retry(self, mock_do_fetch_with_retry):
+        client_with_retry = DefaultCmabClient(
+            http_client=self.mock_http_client,
+            logger=self.mock_logger,
+            retry_config=self.retry_config
+        )
+        result = client_with_retry.fetch_decision(self.rule_id, self.user_id, self.attributes, self.cmab_uuid)
+        self.assertEqual(result, 'var-xyz')
+        mock_do_fetch_with_retry.assert_called_once()
+
+    @patch.object(DefaultCmabClient, '_do_fetch', side_effect=RequestException("Network error"))
+    def test_fetch_decision_request_exception(self, mock_do_fetch):
+        result = self.client.fetch_decision(self.rule_id, self.user_id, self.attributes, self.cmab_uuid)
+        self.assertIsNone(result)
+        self.mock_logger.error.assert_called_with(Errors.CMAB_FETCH_FAILED.format("Network error"))
+
+    @patch.object(DefaultCmabClient, '_do_fetch', return_value=None)
+    def test_fetch_decision_invalid_response(self, mock_do_fetch):
+        result = self.client.fetch_decision(self.rule_id, self.user_id, self.attributes, self.cmab_uuid)
+        self.assertIsNone(result)
+        self.mock_logger.error.assert_called_once()
```

Comment:
in tests I'd add validating raised exceptions that you may add per my comments?.